### DISCUSSION TITLE
[JS] KeyDown 과 KeyUp

### DISCUSSION THREAD

**Discussion by WooGie911** [wanted-9th-3team/frontend-study#80] [프론트 기술 질문 공간]
UI 이벤트 KeyDown 과 KeyUp의 차이점은 무엇인가요?
만약 두개의 키를 동시에 눌렀을 때만 원하는 이벤트가 발생했으면 한다면 어떻게 해야 할까요?

**Comment by michoball** [comment]
keydown 이벤트는 키보드가 눌렸을때 발생하는 이벤트이고 keyup은 키를 누른 손를 떼면 발생하는 이벤트입니다. 

그래서 keydown은 키를 누르자마자 바로 반응하지만 keyup은 누를 손을 떼지 않는한 발생하지 않습니다. 

두개의 키를 동시에 눌렀을 때만을 원한다면 keyDown이벤트로 어떤 키가 눌렸는지를 듣고, 지정한 두개의 키가 눌린 것을 판단하여 눌렸을 경우에 이벤트를 발생시키면 됩니다. 

지정한 두개의 키가 눌린 것을 판단하기위해서는 눌린 키의 정보를 특정 변수에 저장하고 있으면 됩니다.  Array 나 set, map 등을 이용해서 저장하고 지정한 두개의 키가 모두 들어왔는지를 판단하면 될 것 같습니다.

**Comment by WooGie911** [comment]
이벤트 keydown는 키를 눌렀을 때 발생하고 키를 keyup놓았을 때 발생합니다.
만약 두개의 키를 동시에 눌렀을 때만 원하는 이벤트가 발생하는 코드를 짠다면, addEventListener에 원하는 키들이 입력된다면 원하는 이벤트가 발생되게 코드를 짤 것 같습니다.

예를 들면 Set에 입력된 키들을 입,출 되게 한 뒤 원하는 조건에서 이번트가 발생되게 할 것 같습니다.

ex)

```
function runOnKeys(func, ...codes) {
      let pressed = new Set();

      document.addEventListener('keydown', function(event) {
        pressed.add(event.code);

        for (let code of codes) {
          if (!pressed.has(code)) {
            return;
          }
        }
        pressed.clear();

        func();
      });

      document.addEventListener('keyup', function(event) {
        pressed.delete(event.code);
      });

    }
```

**Comment by javascriptbiggosoo** [comment]
두 개의 키를 동시에 눌렀을 때만 원하는 이벤트가 발생하게 하려면 키보드 이벤트 객체의 `key`나 `code` 속성과 함께 이벤트 핸들러 내에서 두 개의 키가 동시에 눌려있는지를 확인하는 상태를 관리해야 합니다.

```jsx
import React, { useState, useEffect } from 'react';

function Component() {
  const [keys, setKeys] = useState({});

  const downHandler = ({ key }) => {
    setKeys(prevKeys => ({ ...prevKeys, [key]: true }));
  };

  const upHandler = ({ key }) => {
    setKeys(prevKeys => ({ ...prevKeys, [key]: false }));
  };

  useEffect(() => {
    window.addEventListener('keydown', downHandler);
    window.addEventListener('keyup', upHandler);

    return () => {
      window.removeEventListener('keydown', downHandler);
      window.removeEventListener('keyup', upHandler);
    };
  }, []);

  useEffect(() => {
    // 'A'와 'B'키를 동시에 누르는 경우에만 원하는 동작을 수행
    if (keys['a'] && keys['b']) {
      console.log('You pressed both A and B keys!');
    }
  }, [keys]);

  return ...
}

export default Component;
```

이렇게 함으로써 두 개의 키가 동시에 눌러졌는지를 확인할 수 있습니다. 다른 키의 조합을 원한다면, `keys['a'] && keys['b']` 부분을 필요한 키로 변경하면 됩니다. 이 예시에서는 'A' 키와 'B' 키가 동시에 눌러지면 콘솔에 메시지를 출력합니다.
